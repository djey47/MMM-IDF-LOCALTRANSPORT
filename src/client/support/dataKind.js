/* @flow */

export const DATA_TRAFFIC = 'TRAFFIC';