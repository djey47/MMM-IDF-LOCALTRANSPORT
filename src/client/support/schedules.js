/* @flow */

/**
 * @returns entries reduced to single one per destination; arrival times are stored in an array; station ids are removed. 
 */
export default function reduceByDestination(scheduleEntries: Object): Object {
  const processedEntries = {};

  Object.keys(scheduleEntries).forEach(entryKey => {
    const { data: { schedules }, stop } = scheduleEntries[entryKey];      
    const reducedSchedules = [];

    let lastDestination = null;
    let currentSchedule = {};
    schedules.forEach(schedule => {
      const { destination, time } = schedule;
      if (destination !== lastDestination) {
        currentSchedule = { ...schedule, times: [] };
        reducedSchedules.push (currentSchedule);          
      }
      currentSchedule.times.push(time);
      lastDestination = destination;
    });
    
    processedEntries[entryKey] = {
      data: { schedules: reducedSchedules },
      stop,
    };
  });

  return processedEntries;
}
