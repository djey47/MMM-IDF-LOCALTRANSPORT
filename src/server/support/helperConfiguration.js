/* @flow */

import childProcess  from 'child_process';

let cachedNPMConfig = undefined;
/**
 * @return Whole NPM config
 */
export function getNPMConfig() {
  if (cachedNPMConfig) { return cachedNPMConfig; }

  const npmConfigOutput = childProcess.execSync('npm config list --json').toString();
  cachedNPMConfig = JSON.parse(npmConfigOutput);

  return cachedNPMConfig;
}
