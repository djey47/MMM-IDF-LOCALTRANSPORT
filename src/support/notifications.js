/* @flow */

/**
 * Notification constants. To be shared between client and server sides.
 */

export const NOTIF_TRANSPORT = 'TRANSPORT';

export const NOTIF_VELIB = 'VELIB';

export const NOTIF_TRAFFIC = 'TRAFFIC';

export const NOTIF_INIT = 'INIT';

export const NOTIF_UPDATE = 'UPDATE';

export const NOTIF_SET_CONFIG = 'SET_CONFIG';

/**
 * MM2 Internal notifications
 */

export const  NOTIF_DOM_OBJECTS_CREATED = 'DOM_OBJECTS_CREATED';
