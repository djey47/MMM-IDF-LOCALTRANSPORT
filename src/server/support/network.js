/* @flow */

import { getNPMConfig } from './helperConfiguration';

/** Network helper functions */

type ProxySettings = {
  host: string,
  port: string,
  proxyAuth: string,
};

/**
* @returns proxy settings from NPM configuration
*/
export function getProxySettings(): ?ProxySettings {
  const { proxy } = getNPMConfig();
  if (!proxy) { return undefined; }

  const [authPart, hostPart] = proxy.split('@');
  const [, proxyAuth] = authPart.split('//');
  const [host, rawPort] = hostPart.split(':');
  const [port] = rawPort.split('/');

  return {
    host,
    port,
    proxyAuth,
  };
}
