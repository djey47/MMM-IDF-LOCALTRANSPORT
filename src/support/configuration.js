/* @flow */

import { getAllStationInfo } from './railwayRepository';
import { NOTIF_SET_CONFIG } from './notifications';

import { resolveIndexFromStopConfig } from './api/api';

import type { NotificationSenderFunction } from '../types/Application';
import type { ModuleConfiguration, StationConfiguration } from '../types/Configuration';
import type { StationInfoResult } from '../types/Transport';

/**
 * Official module name
 */
export const MODULE_NAME = 'MMM-IDF-LOCALTRANSPORT';

/**
 * Id of wrapper DOM node
 */
export const WRAPPER_ID = 'IDFTransportWrapper';

/**
 * All stop configuration types
 */
export const TYPE_BUS = 'bus';
export const TYPE_METRO = 'metros';
export const TYPE_RER = 'rers';
export const TYPE_TRAMWAY = 'tramways';
export const TYPE_TRANSILIEN = 'transiliens';
export const TYPE_TRAFFIC_LEGACY = 'traffic';
export const TYPE_TRAFFIC_TRANSILIEN = 'transiliensTraffic';
export const TYPE_VELIB = 'velib';

/**
 * Default configuration
 */
export const defaults: ModuleConfiguration = {
  maximumEntries: 2,
  maxTimeOffset: 200,
  updateInterval: 1 * 60 * 1000,
  convertToWaitingTime: true,
  initialLoadDelay: 0,
  maxLettersForDestination: 22,
  concatenateArrivals: true,
  showSecondsToNextUpdate: true,
  showLastUpdateTime: false,
  oldUpdateOpacity: 0.5,
  oldThreshold: 0.1,
  debug: false,
  velibGraphWidth: 400,
  velibTrendWidth: 400,
  velibTrendHeight: 100,
  velibTrendTimeScale: 3600,
  velibTrendZoom: 1800,
  velibTrendDay: false,
  trendGraphOff: false,

  apiBaseV3: 'https://api-ratp.pierre-grimaud.fr/v3/',
  apiTransilien: 'http://api.transilien.com/',
  apiVelib: 'https://opendata.paris.fr/api/records/1.0/search/?dataset=stations-velib-disponibilites-en-temps-reel',
  apiAutolib: 'https://opendata.paris.fr/explore/dataset/stations_et_espaces_autolib_de_la_metropole_parisienne/api/',
  apiSncfData: 'https://ressources.data.sncf.com/api/records/1.0/',
  apiCitymapper: 'https://citymapper.com/api/1/',
  transilienToken: 'Basic',
  citymapperToken: '00000000000000000000000000000000',
  messages: {
    ago: 'ago',
    loading: 'Loading connections ...',
    notYet: 'no info yet',
    nextUpdate: 'next update in',
    requestedUpdate: 'update requested',
    unavailable: '-',
    theorical: '?',
    status: {
      approaching: 'Approaching',
      atplatform: 'At platform',
      atstop: 'At stop',
      ontime: 'üòä‚è≤',
      deleted: 'üòû‚ùå',
      delayed: 'üòê‚è≥',
      skipped: '‚ùå',
      terminal: '‚ùå term',
    },
    traffic: {
      ok: 'üòä',
      okwork: 'üòê',
      ko: 'üòû',
    },
    units: {
      minutes: 'mn',
      seconds: 's',
    },
    velib: {
      bikes: 'velibs',
      spaces: 'spaces',
    },
  },

  stations: [],

  blockOrder: {
    traffic: 1,
    schedules: 2,
    velib: 3,
  },   
};

/**
 * Callback to handle async response.
 * Exported for testing.
 * @param {Array<Object>} responses 
 * @param {Function} sendSocketNotification 
 * @param {Object} configuration 
 */
export function handleStationInfoResponse(responses: Array<StationInfoResult>, sendSocketNotification: NotificationSenderFunction, configuration: ModuleConfiguration) {
  const { debug } = configuration;

  responses.forEach(response => {
    if (debug) {
      console.log('** getAllStationInfo response:');
      console.dir(response);
    }

    const { index, stationInfo, destinationInfo } = response;
    configuration.stations[index].uic = {
      station: stationInfo.code_uic,
      destination: destinationInfo ? destinationInfo.code_uic : null,
    };

    // TODO use MM2 logger (available in client context ?)
    const { station, destination, uic } = configuration.stations[index];
    if (uic) {
      console.log(`** ${MODULE_NAME}: Configuration resolved UIC codes:`);
      console.log(`station: ${station || '/'} => ${uic.station || '/'} , destination: ${destination || '/'} => ${uic.destination || '/'}`);
    } else {
      console.error(`** ${MODULE_NAME}: Configuration resolved no UIC codes:`);      
      console.error(`station: ${station || '/'}, destination: ${destination || '/'}`);
    }
  });

  sendSocketNotification(NOTIF_SET_CONFIG, configuration);
}

/**
 * Resolves useful information from module configuration (station UIC ...)
 * Sends configuration to server-side via sockets.
 * @param {Object} configuration configuration to be enhanced
 * @param {Function} sendSocketNotification callback to notification handler
 */
export function enhanceConfiguration(configuration: ModuleConfiguration, sendSocketNotification: (notification: string, payload: Object) => void) {
  const { stations } = configuration;
  // Stations for transilien: retrieve UIC
  const queries = stations
    .filter(stationConfig => stationConfig.type === TYPE_TRANSILIEN)
    .filter(stationConfig => {
      // Do not resolve to UIC codes if already provided
      const { destination, uic } = stationConfig;      
      return !uic || !uic.station || destination && !uic.destination;
    })
    .map((stationConfig, index)  => {
      const { station, destination } = stationConfig;
      
      if(!station) {
        console.error(`** ${MODULE_NAME}: Configuration does not contain station:`);
        console.error(stationConfig);
      }
      
      return {
        index,
        stationValue: station || '?',
        destinationValue: destination,
      };
    });

  if (queries.length) {
    getAllStationInfo(queries, configuration)
      .then(responses => handleStationInfoResponse(responses, sendSocketNotification, configuration));
  } else {
    sendSocketNotification(NOTIF_SET_CONFIG, configuration);
  }
}

/**
 * @returns StationConfiguration with provided identifier if it exists, undefined otherwise.
 */
export function fetchStopConfiguration(config: ModuleConfiguration, id: string): StationConfiguration {
  const { stations } = config;
  const result = stations.filter(station => resolveIndexFromStopConfig(station) === id);
  return result[0];
}
