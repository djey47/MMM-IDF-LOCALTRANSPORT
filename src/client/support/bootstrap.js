/* @flow */

import React from 'react';
import ReactDOM from 'react-dom';

import type { ModuleConfiguration } from '../../types/Configuration';

import Main from '../components/Main/Main';
import { WRAPPER_ID } from '../../support/configuration';

// TODO add css loader to webpack
// import '../../../css/module.css';

/** REACT init helper */
const ModuleBootstrap = (config: ModuleConfiguration) => {
  ReactDOM.render(
    <Main config={config} />,
    document.getElementById(WRAPPER_ID)
  );
};

export default ModuleBootstrap;
