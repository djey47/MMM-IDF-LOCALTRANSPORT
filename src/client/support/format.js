/* @flow */

import moment from 'moment-timezone';
import { translate, MessageKeys } from '../../support/messages';

import type Moment from 'moment';

/**
 * @returns time with format H:MM:SS from moment-compliant string (YYYYMMDDTHHMMSS, ISO...)
 */
export const toHoursMinutesSeconds = (date?: string): string => {
  if (!date) return '';

  return moment(date).format('HH:mm:ss');
};

/**
 * @return time with format HH:MM from moment-compliant string (YYYYMMDDTHHMMSS, ISO...)
 */
export const toHoursMinutes = (date?: string): string => {
  if (!date) return '';

  return moment(date).format('HH:mm');
};

/**
 * @return waiting time in minutes between startTime and transport time ISO
 */
export const toWaitingTime = (transportTime?: string, startTime: Moment, messages: Object): string => {
  const endDate = moment(transportTime);
  if (!endDate.isValid()) return transportTime || '';

  let waitingTime = Math.floor(endDate.diff(startTime) / 1000 / 60);
  return `${waitingTime > 0 ? waitingTime : 0} ${translate(MessageKeys.UNITS_MINUTES, messages)}`;
};
