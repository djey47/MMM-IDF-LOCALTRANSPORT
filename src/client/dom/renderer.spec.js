/* @flow */

import moment from 'moment-timezone';
import htmlBeautify from 'html-beautify';
import {
  renderWrapper,
  renderHeader,
  renderTrafficLegacy,
  renderTrafficTransilien,
  renderPublicTransport,
  renderNoInfoVelib,
  renderVelib,
} from './renderer.js';
import { defaults } from '../../support/configuration';

import type { ModuleConfiguration } from '../../types/Configuration';

const mockNow = jest.fn();
jest.mock('../support/date', () => ({
  now: () => mockNow(),
}));

beforeAll(() => {
  moment.tz.setDefault('UTC');
  mockNow.mockImplementation(() => moment());

  // Fixes JSDOM lib error, see https://github.com/jsdom/jsdom/issues/1782
  window.HTMLCanvasElement.prototype.getContext = () => (null);
});

const testRender = (element: any): String => {
  if (element instanceof Array) {
    const parentNode = document.createElement('table');
    element.forEach(subElement => parentNode.appendChild(subElement));
    return htmlBeautify(parentNode.outerHTML);
  }

  return htmlBeautify(element.outerHTML);
};

describe('renderWrapper function', () => {
  it('should return correct HTML when not loaded', () => {
    // given-when
    const actual = renderWrapper(false);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when loaded', () => {
    // given-when
    const actual = renderWrapper(true);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });
});

describe('renderHeader function', () => {
  const baseConfig = {
    ...defaults,
    messages: {},
    lastUpdate: moment('2017-07-27T08:23:40.000Z'),
    showLastUpdateTime: false,
    showSecondsToNextUpdate: false,
  };
  const data = {
    header: 'Connections',
  };

  it('should return correct header when complete configuration', () => {
    // given
    mockNow.mockImplementation(() => moment('2017-07-27T08:23:55Z'));
    const config: ModuleConfiguration = { ...baseConfig, showLastUpdateTime: true, showSecondsToNextUpdate: true };
    // when
    const actual = renderHeader(data, config);
    // then
    expect(actual).toMatchSnapshot();
  });

  it('should return correct header when incomplete configuration 1', () => {
    // given
    mockNow.mockImplementation(() => moment('2017-07-27T08:23:55Z'));
    const config: ModuleConfiguration = { ...baseConfig, showSecondsToNextUpdate: true };
    // when
    const actual = renderHeader(data, config);
    // then
    expect(actual).toMatchSnapshot();
  });

  it('should return correct header when incomplete configuration 2', () => {
    // given
    mockNow.mockImplementation(() => moment('2017-07-27T08:23:55Z'));
    const config: ModuleConfiguration = { ...baseConfig, showLastUpdateTime: true };
    // when
    const actual = renderHeader(data, config);
    // then
    expect(actual).toMatchSnapshot();
  });

  it('should return simple string when silent configuration', () => {
    // given
    mockNow.mockImplementation(() => moment('2017-07-27T08:23:55Z'));
    const config: ModuleConfiguration = { ...baseConfig, showLastUpdateTime: false };
    // when
    const actual = renderHeader(data, config);
    // then
    expect(actual).toMatchSnapshot();
  });
});

describe('renderPublicTransport function', () => {
  mockNow.mockImplementation(() => moment('2017-05-30T12:45:00Z'));
  const stop = {
    line: ['BUS', 275],
    station: 'Ulbach',
    destination: 'La+Defense',
  };
  const stopIndex = 'bus,275/Ulbach/La+Defense';
  const baseConfig = {
    ...defaults,
    convertToWaitingTime: false,
    maxLettersForDestination: 256,
    maximumEntries: 3,
    messages: {},
  };

  it('should return correct HTML when no schedule', () => {
    // given
    const stop = {
      line: ['BUS', 275],
      station: '',
    };
    const schedules = {};
    const config = {
      ...defaults,
      maximumEntries: 1,
      messages: {},
    };
    // when
    const actual = renderPublicTransport(stop, stopIndex, schedules, {}, config);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when schedule', () => {
    // given
    const schedules = {
      [stopIndex]: [{
        time: '2017-07-16T13:00:00.000Z',
        destination: 'La Défense',
      },{
        time: '2017-07-16T13:05:00.000Z',
        destination: 'Place Charras',
      }],
    };
    const lastUpdate = {
      [stopIndex]: '2017-05-30T15:00:00.000Z',
    };
    // when
    const actual = renderPublicTransport(stop, stopIndex, schedules, lastUpdate, baseConfig);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when schedule and status info', () => {
    // given
    const rerStop = {
      line: 'A',
      station: 'La+Defense',
      destination: 'a',
    };
    const rerStopIndex = 'a/la+defense/a';
    const schedules = {
      [rerStopIndex]: [{
        time: '2017-05-30T13:00:00.000Z',
        status: 'DELAYED',
        destination: 'La Défense',
        code: 'UAPY',
      }],
    };
    const lastUpdate = {
      [rerStopIndex]: '2017-05-30T15:00:00.000Z',
    };
    // when
    const actual = renderPublicTransport(rerStop, rerStopIndex, schedules, lastUpdate, baseConfig);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when schedule and mission code', () => {
    // given
    const schedules = {
      [stopIndex]: [{
        time: '2017-05-30T13:00:00.000Z',
        status: '',
        code: 'POPU',
        destination: 'La Défense',
      }],
    };
    const lastUpdate = {
      [stopIndex]: '2017-05-30T15:00:00.000Z',
    };
    // when
    const actual = renderPublicTransport(stop, stopIndex, schedules, lastUpdate, baseConfig);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when schedule and theorical time mode', () => {
    // given
    const schedules = {
      [stopIndex]: [{
        time: '2017-05-30T13:00:00.000Z',
        timeMode: 'THEORICAL',
        destination: 'La Défense',
      }],
    };
    const lastUpdate = {
      [stopIndex]: '2017-05-30T15:00:00.000Z',
    };
    // when
    const actual = renderPublicTransport(stop, stopIndex, schedules, lastUpdate, baseConfig);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when schedule and convert to waiting time', () => {
    // given
    mockNow.mockImplementation(() => moment('2017-05-30T12:45:00.000Z'));
    const schedules = {
      [stopIndex]: [{
        time: '2017-05-30T13:00:00.000Z',
        destination: 'La Défense',
      },{
        time: '2017-05-30T13:15:00.000Z',
        destination: 'Place Charras',
      }],
    };
    const lastUpdate = {
      [stopIndex]: '2017-05-30T15:00:00.000Z',
    };
    const config = {
      ...baseConfig,
      convertToWaitingTime: true,
    };
    // when
    const actual = renderPublicTransport(stop, stopIndex, schedules, lastUpdate, config);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when schedule and concatenate arrivals', () => {
    // given
    const schedules = {
      [stopIndex]: [{
        time: '2017-07-16T13:00:00.000Z',
        destination: 'Place Charras',
      },{
        time: '2017-07-16T13:15:00.000Z',
        destination: 'La Défense',
      },{
        time: '2017-07-16T13:30:00.000Z',
        destination: 'La Défense',
      }],
    };
    const lastUpdate = {
      [stopIndex]: '2017-05-30T15:00:00.000Z',
    };
    const config = {
      ...baseConfig,
      concatenateArrivals: true,
    };
    // when
    const actual = renderPublicTransport(stop, stopIndex, schedules, lastUpdate, config);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when schedules and concatenate arrivals for transilien', () => {
    // given
    const stopIndexTransilien = 'gare/87382002/depart';
    const stopConfigTransilien = {
      type: 'transiliens',
      label: 'Becon',
      station: 'Becon Les Bruyeres',
      destination: 'Saint Nom La breteche',
      uic: {
        station: '87382002',
        destination: '87382481',
      },
    };
    const schedulesTransilien = {
      [stopIndexTransilien]: [{
        destination: 'SAINT-NOM LA BRETECHE FORET DE MARLY',
        code: 'SEBU',
        time: '2017-07-26T11:20:00.000Z',
        status: 'ON_TIME',
      },
      {
        destination: 'SAINT-NOM LA BRETECHE FORET DE MARLY',
        code: 'SEBU',
        time: '2017-07-26T11:35:00.000Z',
        status: 'DELAYED',
      },
      {
        destination: 'SAINT-NOM LA BRETECHE FORET DE MARLY',
        code: 'SEBU',
        time: '2017-07-26T11:50:00.000Z',
        status: 'DELETED',
      },
      {
        destination: 'SAINT-NOM LA BRETECHE FORET DE MARLY',
        code: 'SEBU',
        time: '2017-07-26T12:05:00.000Z',
        status: 'UNKNOWN',
      },
      {
        destination: 'SAINT-NOM LA BRETECHE FORET DE MARLY',
        code: 'SEBU',
        time: '2017-07-26T12:20:00.000Z',
      }],
    };
    const lastUpdateTransilien = {
      [stopIndexTransilien]: '2017-07-26T13:17:00.000Z',
    };
    const config = {
      ...baseConfig,
      concatenateArrivals: true,
    };
    // when
    const actual = renderPublicTransport(stopConfigTransilien, stopIndexTransilien, schedulesTransilien, lastUpdateTransilien, config);
    // then
    expect(actual.length).toEqual(1);
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when no schedule and no custom label for transilien', () => {
    // given
    const stopConfigTransilien = {
      type: 'transiliens',
      station: 'Becon Les Bruyeres',
      destination: 'Saint Nom La breteche',
      uic: {
        station: '87382002',
        destination: '87382481',
      },
    };
    const config = {
      ...baseConfig,
      concatenateArrivals: true,
    };
    // when
    const actual = renderPublicTransport(stopConfigTransilien, 'gare/87382002/depart', {}, {}, config);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });
});

describe('renderTrafficLegacy function', () => {
  it('should return correct HTML for table row', () => {
    // given
    const stop = {
      type: 'traffic',
      line: ['BUS', 275],
      label: 'Ulbach Label',
    };
    const ratpTraffic = {
      'traffic/bus/275': {
        status: 'OK_WORK',
        message: 'message',
        summary: 'summary',
      },
    };
    const config = { ...defaults, messages: {} };
    // when
    const actual = renderTrafficLegacy(stop, ratpTraffic, config);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });
});

describe('renderTrafficTransilien function', () => {
  it('should return correct HTML for table row when message', () => {
    // given
    const stop = {
      type: 'transiliensTraffic',
      line: 'L',
      label: 'L Label',
    };
    const transilienTraffic = {
      'traffic/transiliens/l': {
        status: 'UNKNOWN',
        message: 'message',
        summary: 'summary',
      },
    };
    const config = { ...defaults, messages: {} };
    // when
    const actual = renderTrafficTransilien(stop, transilienTraffic, config);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML for table row when no message', () => {
    // given
    const stop = {
      type: 'transiliensTraffic',
      line: 'L',
      label: 'L Label',
    };
    const transilienTraffic = {
      'traffic/transiliens/l': {
        status: 'UNKNOWN',
        summary: 'summary',
      },
    };
    const config = { ...defaults, messages: {} };
    // when
    const actual = renderTrafficTransilien(stop, transilienTraffic, config);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });
});

const baseStopConfigForVelib = {
  type: 'velibs',
  line: ['VELIB', 68],
};

describe('renderNoInfoVelib function', () => {
  it('should return correct HTML for table row when label', () => {
    // given
    const stop = {
      ...baseStopConfigForVelib,
      label: 'Ulbach',
      station: '',
    };
    // when
    const actual = renderNoInfoVelib(stop);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML for table cell when no label', () => {
    // given
    const stop = {
      ...baseStopConfigForVelib,
      station: 'Stations',
    };
    // when
    const actual = renderNoInfoVelib(stop);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });
});

describe('renderVelib function', () => {
  it('should return correct HTML when no history', () => {
    // given
    const stop = {
      ...baseStopConfigForVelib,
      label: 'Opera Bis',
      station: '',
    };
    const velibHistory = {};
    const config = {
      ...defaults,
      messages: {},
    };
    // when
    const actual = renderVelib(stop, velibHistory, config);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when history without trend', () => {
    // given
    const stop = {
      ...baseStopConfigForVelib,
      station: '2209',
    };
    const velibHistory = {
      '2209': [{
        total: 10,
        bike: 2,
        empty: 8,
        name: 'Opera',
      }],
    };
    const config = {
      ...defaults,
      trendGraphOff: true,
      messages: {},
    };
    mockNow.mockImplementationOnce(() => new Date(2017, 5, 29, 8, 34, 28));
    // when
    const actual = renderVelib(stop, velibHistory, config);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });

  it('should return correct HTML when history with trend', () => {
    // given
    const stop = {
      ...baseStopConfigForVelib,
      station: '2209',
    };
    const velibHistory = {
      '2209': [{
        total: 10,
        bike: 2,
        empty: 8,
        name: 'Opera',
      }],
    };
    const config = { ...defaults };
    mockNow.mockImplementationOnce(() => new Date(2017, 5, 29, 8, 34, 28));
    // when
    const actual = renderVelib(stop, velibHistory, config);
    // then
    expect(testRender(actual)).toMatchSnapshot();
  });
});
