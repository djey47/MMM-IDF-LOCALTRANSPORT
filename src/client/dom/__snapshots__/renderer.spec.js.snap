// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderHeader function should return correct header when complete configuration 1`] = `"Connections, {nextUpdate} 45{units.seconds} @ 10:23:40"`;

exports[`renderHeader function should return correct header when incomplete configuration 1 1`] = `"Connections, {nextUpdate} 45{units.seconds}"`;

exports[`renderHeader function should return correct header when incomplete configuration 2 1`] = `"Connections @ 10:23:40"`;

exports[`renderHeader function should return simple string when silent configuration 1`] = `"Connections"`;

exports[`renderNoInfoVelib function should return correct HTML for table cell when no label 1`] = `
"<tr>
   <td class=\\"bright\\">
      Stations {notYet}
   </td>
</tr>"
`;

exports[`renderNoInfoVelib function should return correct HTML for table row when label 1`] = `
"<tr>
   <td class=\\"bright\\">
      Ulbach {notYet}
   </td>
</tr>"
`;

exports[`renderPublicTransport function should return correct HTML when no schedule 1`] = `
"<table>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\">
         BUS,275
      </td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         {unavailable}
      </td>
      <td></td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         {unavailable}
      </td>
</tr>
</table>"
`;

exports[`renderPublicTransport function should return correct HTML when no schedule and no custom label for transilien 1`] = `
"<table>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\">
         Becon Les Bruyeres
      </td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         {unavailable}
      </td>
      <td></td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         {unavailable}
      </td>
</tr>
</table>"
`;

exports[`renderPublicTransport function should return correct HTML when schedule 1`] = `
"<table>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\">
         BUS,275
      </td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         La Défense
      </td>
      <td></td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         13:00
      </td>
</tr>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\"></td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         Place Charras
      </td>
      <td></td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         13:05
      </td>
</tr>
</table>"
`;

exports[`renderPublicTransport function should return correct HTML when schedule and concatenate arrivals 1`] = `
"<table>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\">
         BUS,275
      </td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         Place Charras
      </td>
      <td></td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         13:00
      </td>
</tr>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\"></td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         La Défense
      </td>
      <td></td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         13:15 / 13:30
      </td>
</tr>
</table>"
`;

exports[`renderPublicTransport function should return correct HTML when schedule and convert to waiting time 1`] = `
"<table>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\">
         BUS,275
      </td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         La Défense
      </td>
      <td></td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         15 {units.minutes}
      </td>
</tr>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\"></td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         Place Charras
      </td>
      <td></td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         30 {units.minutes}
      </td>
</tr>
</table>"
`;

exports[`renderPublicTransport function should return correct HTML when schedule and mission code 1`] = `
"<table>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\">
         BUS,275
      </td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         La Défense
      </td>
      <td>
         POPU
      </td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         13:00
      </td>
</tr>
</table>"
`;

exports[`renderPublicTransport function should return correct HTML when schedule and status info 1`] = `
"<table>
   <tr class=\\"Schedules__item bright is-delayed\\">
      <td class=\\"align-right\\">
         A
      </td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         La Défense
      </td>
      <td>
         UAPY {status.delayed}
      </td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         13:00
      </td>
</tr>
</table>"
`;

exports[`renderPublicTransport function should return correct HTML when schedule and theorical time mode 1`] = `
"<table>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\">
         BUS,275
      </td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         La Défense
      </td>
      <td></td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         13:00{theorical}
      </td>
</tr>
</table>"
`;

exports[`renderPublicTransport function should return correct HTML when schedules and concatenate arrivals for transilien 1`] = `
"<table>
   <tr class=\\"Schedules__item bright\\">
      <td class=\\"align-right\\">
         Becon
      </td>
      <td class=\\"align-left\\" style=\\"opacity: 0.5;\\">
         SAINT-NOM LA BRETECHE FORET DE MARLY
      </td>
      <td>
         SEBU
      </td>
      <td class=\\"\\" style=\\"opacity: 0.5;\\">
         11:20 / 11:35 / 11:50
      </td>
</tr>
</table>"
`;

exports[`renderTrafficLegacy function should return correct HTML for table row 1`] = `
"<tr class=\\"Traffic__item bright is-ok-with-work\\">
   <td class=\\"align-right\\">
      Ulbach Label
   </td>
   <td>
      {traffic.okwork}
   </td>
   <td class=\\"align-left\\" colspan=\\"2\\">
      <span class=\\"Traffic__title\\">summary</span>
      <div class=\\"Traffic__messageContainer\\">
         <div class=\\"Traffic__messageContents\\">
            message
         </div>
</div>
</td>
</tr>"
`;

exports[`renderTrafficTransilien function should return correct HTML for table row when message 1`] = `
"<tr class=\\"Traffic__item bright\\">
   <td class=\\"align-right\\">
      L Label
   </td>
   <td>
      {unavailable}
   </td>
   <td class=\\"align-left\\" colspan=\\"2\\">
      <span class=\\"Traffic__title\\">summary</span>
      <div class=\\"Traffic__messageContainer\\">
         <div class=\\"Traffic__messageContents\\">
            message
         </div>
</div>
</td>
</tr>"
`;

exports[`renderTrafficTransilien function should return correct HTML for table row when no message 1`] = `
"<tr class=\\"Traffic__item bright\\">
   <td class=\\"align-right\\">
      L Label
   </td>
   <td>
      {unavailable}
   </td>
   <td class=\\"align-left\\" colspan=\\"2\\">
      <span class=\\"Traffic__title\\">summary</span>
      <div class=\\"Traffic__messageContainer\\">
         <div class=\\"Traffic__messageContents\\"></div>
</div>
</td>
</tr>"
`;

exports[`renderVelib function should return correct HTML when history with trend 1`] = `"<tr></tr>"`;

exports[`renderVelib function should return correct HTML when history without trend 1`] = `
"<tr>
   <td class=\\"align-left\\">
      10
   </td>
   <td class=\\"bright\\">
      2 {velib.bikes} 8 {velib.spaces}
   </td>
   <td class=\\"align-right\\">
      Opera
   </td>
   <td></td>
</tr>"
`;

exports[`renderVelib function should return correct HTML when no history 1`] = `
"<tr>
   <td class=\\"bright\\">
      Opera Bis {notYet}
   </td>
</tr>"
`;

exports[`renderWrapper function should return correct HTML 1`] = `"<div id=\\"IDFTransportWrapper\\" class=\\"IDFTransportWrapper\\"></div>"`;
